src/batSet.ml

```diff
@@ -458,6 +555,57 @@ module Concrete = struct
         join (union cmp12 l1 l2) v2 (union cmp12 r1 r2)
       end
 
+  let rec filter p = function
+      Empty -> Empty
+    | (Node(l,v,r,_)) as t ->
+       (* call [p] in the expected left-to-right order *)
+       let l' = filter p l in
+       let pv = p v in
+       let r' = filter p r in
+       if pv then
+         if l==l' && r==r' then t else join l' v r'
+       else concat l' r'
+                       
+  let try_join cmp l v r =
+    (* [join l v r] can only be called when (elements of l < v <
+         elements of r); use [try_join l v r] when this property may
+         not hold, but you hope it does hold in the common case *)
+    if (l = Empty || cmp (max_elt l) v < 0)
+       && (r = Empty || cmp v (min_elt r) < 0)
+    then join l v r
+    else union cmp l (add cmp v r)
+
+  let rec map_stdlib cmp f = function
```

Comment:
the map and filter_map functions from stdlib take a function elt -> elt, while the functions in Batteries take a function elt -> 'b and produce a 'b set (for the polymorphic maps)
So it seems we must keep the old versions or break compatibility with people that currently rely on map/filter_map to create sets of a different element type.

Reply:
Ah yes, thanks. I would propose to call them `map_endo` or `endo_map` (from "endo{morphism,function}", a function from a set to itself) or `homogeneous_map` rather than `map_stdlib`.

The implementation of `map_stdlib` and `filter_map_stdlib` that you propose is more efficient than the existing `map` and `filter_map` functions in the case where the mapping function preserves the order, or "nearly" preserves the order. Would you possibly be interested in using the same implementation with `try_join` for the existing/heterogeneous map functions? Note that the `==` trick cannot be reused, as it assumes that the input and output types are equal.

Reply:
To give credit where it is due: the map_stdlib and filter_map_stdlib methods are copied (mostly) verbatim from the stdlib.

I rewrote the map and filter_map functions to use try_join. I'm not convinced that makes them faster though...